syntax = "proto3";
option go_package = ".;gRPC";

service Algorithm {
  rpc Ping(PingRequest) returns (PingReply) {}
  rpc Schedule(ScheduleRequest) returns (ScheduleReply) {}
}

message PingRequest { string message = 1; }

message PingReply { string message = 1; }

message ScheduleRequest {
  ItemList request = 1;
  uint32 num_deliverer = 2;
  repeated ItemList itemlists = 3;
  // id: 0 purchaser, 1 - n supplier, n+1 - m deliverer
  // (0, 1) (0, 2) ... (0, n) (1, 1) (1, 2)... (1, m) (2, 1) (2, 2) ... (2, m) ... (n, m)
  repeated uint64 distance = 4;
}

message ScheduleReply {
  // 1 based index
  uint32 deliverer_id = 1;
  repeated Route route = 2;
}

message ItemList { map<uint32, double> items = 1; }

message Route {
  // 1 based index
  uint32 supplier_id = 1;
  ItemList itemlist = 2;
}
